source/h2_stream.c

```diff
@@ -524,6 +559,34 @@ int aws_h2_stream_on_decoder_data(struct aws_h2_stream *stream, struct aws_byte_
     return AWS_OP_SUCCESS;
 }
 
+int aws_h2_stream_on_decoder_window_update(struct aws_h2_stream *stream, uint32_t window_size_increment) {
+    AWS_PRECONDITION_ON_CHANNEL_THREAD(stream);
+
+    if (s_check_state_allows_frame_type(stream, AWS_H2_FRAME_T_WINDOW_UPDATE)) {
+        return s_send_rst_and_close_stream(stream, aws_last_error());
+    }
+    if (window_size_increment == 0) {
+        /* flow-control winodw increment of 0 MUST be treated as error (RFC7540 6.9.1) */
+        AWS_H2_STREAM_LOG(ERROR, stream, "Window udpate frame with 0 increment size");
+        return s_send_rst_and_close_stream(stream, AWS_ERROR_HTTP_PROTOCOL_ERROR);
+    }
+    int32_t old_window_size = stream->thread_data.window_size_peer;
+    if (aws_h2_stream_window_size_change(stream, window_size_increment)) {
+        /* We MUST NOT allow a flow-control window to exceed the max */
+        AWS_H2_STREAM_LOG(
+            ERROR, stream, "Window udpate frame causes the connection flow-control window exceeding the maximum size");
+        return s_send_rst_and_close_stream(stream, AWS_ERROR_HTTP_FLOW_CONTROL_ERROR);
+    }
+    if (stream->thread_data.window_size_peer > 0 && old_window_size <= 0) {
+        struct aws_h2_connection *connection = s_get_h2_connection(stream);
+        /* It may already be in the outgoing stream list, but we can remove it and put it at the back again, for
+         * simplicity */
+        aws_linked_list_remove(&stream->node);
+        aws_linked_list_push_back(&connection->thread_data.outgoing_streams_list, &stream->node);
+    }
```

Comment:
it's weird for the stream to be manipulating the connection's datastructures.
maybe have an out-param like `window_flipped_positive` or `stream_unstalled` that gets set true, and the connection can move the stream around based on that

that's what I did with:
`int aws_h2_stream_on_activated(struct aws_h2_stream *stream, bool *out_has_outgoing_data);`
the out-param tells the connection whether or not to put the stream in `outgoing_streams_list`